libcore/FileOperations/CopyMoveJob.vala

```diff
@@ -34,12 +34,211 @@ public class Files.FileOperations.CopyMoveJob : CommonJob {
         base (parent_window);
         this.files = files.copy_deep ((GLib.CopyFunc<GLib.File>) GLib.Object.ref);
         this.destination = destination;
+
+        undo_redo_data = new UndoActionData (Files.UndoActionType.COPY, files.length ());
+        undo_redo_data.set_src_dir (files.data.get_parent ());
+        undo_redo_data.set_dest_dir (destination);
     }
 
     public CopyMoveJob.move (Gtk.Window? parent_window, GLib.List<GLib.File> files, GLib.File? destination) {
         base (parent_window);
         this.files = files.copy_deep ((GLib.CopyFunc<GLib.File>) GLib.Object.ref);
         this.destination = destination;
         is_move = true;
+
+        if (files.data.has_uri_scheme ("trash")) {
+            undo_redo_data = new UndoActionData (Files.UndoActionType.RESTOREFROMTRASH, files.length ());
+        } else {
+            undo_redo_data = new UndoActionData (Files.UndoActionType.MOVE, files.length ());
+        }
+
+        undo_redo_data.set_src_dir (files.data.get_parent ());
+    }
+
+    public CopyMoveJob.link (Gtk.Window? parent_window, GLib.List<GLib.File> files, GLib.File destination) {
+        base (parent_window);
+        this.files = files.copy_deep ((GLib.CopyFunc<GLib.File>) GLib.Object.ref);
+        this.destination = destination;
+
+        undo_redo_data = new UndoActionData (Files.UndoActionType.CREATELINK, files.length ());
+        undo_redo_data.set_src_dir (files.data.get_parent ());
+        undo_redo_data.set_dest_dir (destination);
+    }
+
+    public CopyMoveJob.duplicate (Gtk.Window? parent_window, GLib.List<GLib.File> files) {
+        base (parent_window);
+        this.files = files.copy_deep ((GLib.CopyFunc<GLib.File>) GLib.Object.ref);
+
+        undo_redo_data = new UndoActionData (Files.UndoActionType.DUPLICATE, files.length ());
+        var parent_dir = files.data.get_parent ();
+        undo_redo_data.set_src_dir (parent_dir);
+        undo_redo_data.set_dest_dir (parent_dir);
+    }
+
+    public async bool do_link (GLib.Cancellable? _cancellable) throws GLib.Error {
```

Comment:
Not sure if this needs to be async as it does not seem to call any async functions?

Reply:
It helps to have a direct replacement of the equivalent function that was on C side